DEPR: Deprecate is_copy (#18801)

cbertinato opened the pull request:
Not all tests for this branch on the fork are passing, but these are the same tests that fail for the clean master and appear to be unrelated to the changes.

References to is_copy in tests were changed to the internal attribute _is_copy assuming that, though is_copy is not meant to be a public attribute, checking the state of this attribute is necessary to check consistent internal behavior.

Closes #18801.

gfyoung commented on pandas/core/generic.py:
```diff
@@ -132,10 +132,22 @@ def __init__(self, data, axes=None, copy=False, dtype=None,
+    @property
+    def is_copy(self):
+        warnings.warn("Attribute 'is_copy' will be removed in a future "
+                      "version.", FutureWarning, stacklevel=2)
```
Say : "is deprecated and will removed..."

gfyoung replied:
Also, add a test for this deprecation (both for getter and setter).

jreback reviewed (changes requested):
can you add a whatsnew note in the deprecation section & a test (as @gfyoung indicates); you can put this in pandas/tests/indexing/test_chainging_and_caching.py

Commit:
DEPR: Deprecate is_copy (#18801)

- Renamed 'is_copy' attribute to '_is_copy' for internal use
- Setup getter and setter for 'is_copy'
- Added tests for deprecation warning

pep8speaks commented:
Hello @cbertinato! Thanks for updating the PR.









 - In the file [`pandas/tests/indexing/test_chaining_and_caching.py`](https://github.com/pandas-dev/pandas/blob/9b8f2af0cc434d1e51015ffc77b1715980e12e66/pandas/tests/indexing/test_chaining_and_caching.py), following are the PEP8 issues :

> [Line 424:9](https://github.com/pandas-dev/pandas/blob/9b8f2af0cc434d1e51015ffc77b1715980e12e66/pandas/tests/indexing/test_chaining_and_caching.py#L424): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '

Commit:
Fixed flake8 issues.

cbertinato commented:
Shall I squash these commits into one and resubmit, or can we do it on the merge?

jreback commented:
no need to squash

will be auto done on merge

cbertinato committed:
Merge branch 'master' into issue-18801

cbertinato commented:
I am by no means in a rush, but just wanted to check whether this is waiting on anything else. I believe that the requested changes have been made.

cbertinato commented:
Roger that!

jreback commented:
thanks @cbertinato

topper-123 commented:
When tab-completing in iPython, the ``FutureWarning`` for ``is_copy`` pops up.

I believe that ``'is_copy'`` should be added to ``_deprecations``

jreback commented:
yep that is right - PR?

cbertinato commented:
PR #18922
